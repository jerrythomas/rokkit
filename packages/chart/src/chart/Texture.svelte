<script>
	export let id
	export let fill
	export let path
	export let contrast
	export let thickness = 0.5
	export let patternUnits = 'userSpaceOnUse'
	export let size = 10
</script>

<pattern {id} {patternUnits} width={size} height={size}>
	<rect width={size} height={size} {fill} />
	{#if path}
		<path d={path} fill="none" stroke={contrast} stroke-width={thickness} />
	{/if}
</pattern>
