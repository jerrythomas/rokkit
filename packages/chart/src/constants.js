export const __muted__ = {
	color: '#eeeeee',
	fill: 'empty',
	shape: 'circle'
}
export const __colors__ = [
	'#FFDE6B',
	'#EF89EE',
	'#F79F1E',
	'#02B8FF',
	'#9F84EC',
	'#15CBC4',
	'#0092FD',
	'#F63A57',
	'#A2CB39',
	'#FF6E2F',
	'#FEB8B9',
	'#af7aa1',
	'#7EFFF5'
]

export const __patterns__ = {
	A: 'M0 5A6 6 0 0 0 10 5',
	B: 'M0 10L10 0',
	C: 'M0 0A10 10 0 0 0 10 10',
	D: 'M0 0L10 10',
	E: 'M10 5A6 6 0 0 0 0 5',
	F: 'M10 10A10 10 0 0 0 0 0',
	G: 'M0 0L10 10ZM10 0L0 10Z',
	H: 'M1 1L9 1L9 9L1 9Z',
	I: 'M4 0L4 10M6 10L6 0M0 4L10 4M10 6L0 6',
	J: 'M0 2L8 10M2 0L10 8M0 8L8 0M2 10L10 2',
	K: 'M5 1A 4 4 0 0 0 9 5A4 4 0 0 0 5 9A4 4 0 0 0 1 5A4 4 0 0 0 5 1',
	L: 'M1 3L7 9M3 1L9 7M1 7L7 1M3 9L9 3',
	M: 'M2 2A4 4 0 0 0 8 2A4 4 0 0 0 8 8A4 4 0 0 0 2 8A4 4 0 0 0 2 2',
	N: 'M0 0A5 5 0 0 0 10 0A5 5 0 0 0 10 10A5 5 0 0 0 0 10A5 5 0 0 0 0 0',
	O: 'M5 2A 3 3 0 0 0 8 5A3 3 0 0 0 5 8A3 3 0 0 0 2 5A3 3 0 0 0 5 2',
	P: 'M2 5L5 2L8 5L5 8Z',
	Q: 'M3 5A2 2 0 0 0 7 5A2 2 0 0 0 3 5M1 5L9 5M5 1L5 9',
	R: 'M2 8L8 2ZM1.5 3.5L3.5 1.5ZM6.5 8.5L8.5 6.5ZM0 0L10 10Z',
	S:
		'M2 8L8 2ZM1.5 3.5L3.5 1.5Z' +
		'M6.5 8.5L8.5 6.5Z' +
		'M2 2L8 8M1.5 6.5L3.5 8.5' +
		'M6.5 1.5L8.5 3.5',
	T:
		'M5 1 A6 6 0 0 0 5 9' +
		'A6 6 0 0 0 5 1' +
		'M1 5A6 6 0 0 0 9 5A6 6 0 0 0 1 5',
	U:
		'M1.5 5A1 1 0 0 0 3.5 5A1 1 0 0 0 1.5 5' +
		'M6.5 5A1 1 0 0 0 8.5 5A1 1 0 0 0 6.5 5' +
		'M5 1.5A1 1 0 0 0 5 3.5A1 1 0 0 0 5 1.5' +
		'M5 6.5A1 1 0 0 0 5 8.5A1 1 0 0 0 5 6.5',
	V:
		'M1.5 2.5A1 1 0 0 0 3.5 2.5A1 1 0 0 0 1.5 2.5' +
		'M6.5 2.5A1 1 0 0 0 8.5 2.5A1 1 0 0 0 6.5 2.5' +
		'M2.5 6.5A1 1 0 0 0 2.5 8.5A1 1 0 0 0 2.5 6.5' +
		'M7.5 6.5A1 1 0 0 0 7.5 8.5A1 1 0 0 0 7.5 6.5' +
		'M3.5 5A1 1 0 0 0 6.5 5A1 1 0 0 0 3.5 5',
	W: 'M5 0L6 4L10 5L6 6L5 10L4 6L0 5L4 4Z' + 'M2 1V3M1 2H3' + 'M8 9V7M9 8H7',
	X: 'M5 2L2.5 9L8.8 4.6L1.2 4.6L7.5 9Z',
	Y: 'M0 5A5 5 0 0 0 5 0' + 'M5 10A5 5 0 0 0 0 5' + 'M5 10A5 5 0 0 0 5 0',
	Z: 'M0 0L10 10M5 0L10 5M0 5 L5 10',
	Z1: 'M0 0L10 10M3 0L10 7M0 7 L3 10'
}
