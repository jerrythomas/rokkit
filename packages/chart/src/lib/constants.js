import patterns__ from './patterns.json'
import __palette from './palette.json'

export const __patterns__ = patterns__
export const palette = __palette
export const __muted__ = {
	color: '#eeeeee',
	fill: 'empty',
	shape: 'circle'
}
export const __colors__ = [
	'#FFDE6B',
	'#EF89EE',
	'#F79F1E',
	'#02B8FF',
	'#9F84EC',
	'#15CBC4',
	'#0092FD',
	'#F63A57',
	'#A2CB39',
	'#FF6E2F',
	'#FEB8B9',
	'#af7aa1',
	'#7EFFF5'
]

export const colors = [
	'#FFDE6B',
	'#EF89EE',
	'#F79F1E',
	'#02B8FF',
	'#9F84EC',
	'#15CBC4',
	'#0092FD',
	'#F63A57',
	'#A2CB39',
	'#FF6E2F',
	'#FEB8B9',
	'#af7aa1',
	'#7EFFF5'
]
