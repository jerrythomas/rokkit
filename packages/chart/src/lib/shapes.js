const names = [
	'circle',
	'square',
	'triangle',
	'diamond',
	'star',
	'rhombus',
	'heart'
]
const data = {
	square: ['M', 1, 1, 'L', 9, 1, 'L', 9, 9, 'L', 1, 9, 'Z'],
	circle: [
		['M', 0, 5],
		['A', 5, 5, 0, 0, 0, 10, 5],
		['A', 5, 5, 0, 0, 0, 0, 5]
	],
	triangle: ['M', 5, 0, 'L', 10, 10, 'L', 0, 10, 'Z'],
	diamond: [
		['M', 5, 0],
		['A', 7, 7, 0, 0, 0, 10, 5],
		['A', 7, 7, 0, 0, 0, 5, 10],
		['A', 7, 7, 0, 0, 0, 0, 5],
		['A', 7, 7, 0, 0, 0, 5, 0]
	],
	rhombus: ['M', 0, 5, 'L', 5, 0, 'L', 10, 5, 'L', 5, 10, 'Z'],
	heart: [
		['M', 9, 5],
		['A', 0.8, 0.8, 0, 0, 0, 5, 2],
		['A', 0.8, 0.8, 0, 0, 0, 1, 5],
		['L', 5, 9],
		['L', 9, 5]
	],
	star: [
		['M', 4.80001, 0],
		['L', 5.92258, 3.45491],
		['H', 9.55529],
		['L', 6.61637, 5.59017],
		['L', 7.73894, 9.04509],
		['L', 4.80001, 6.90983],
		['L', 1.86108, 9.04509],
		['L', 2.98365, 5.59017],
		['L', 0.0447266, 3.45491],
		['H', 3.67744],
		['L', 4.80001, 0],
		['Z']
	]
}

function scaledPath(size, x) {
	if (Array.isArray(x)) return x.map((x) => scaledPath(size, x)).join(' ')
	return typeof size === 'number' ? x * size : x
}

export const shapes = names.map((name) => ({
	name,
	path: (s) => scaledPath(s, data[name])
}))

export const namedShapes = {
	square: (s) =>
		`M${0.1 * s} 0` +
		`A${0.1 * s} ${0.1 * s} 0 0 0 0 ${0.1 * s}V${0.9 * s}` +
		`A${0.1 * s} ${0.1 * s} 0 0 0 ${0.1 * s} ${s}H${0.9 * s}` +
		`A${0.1 * s} ${0.1 * s} 0 0 0 ${s} ${0.9 * s}V${0.1 * s}` +
		`A${0.1 * s} ${0.1 * s} 0 0 0 ${0.9 * s} 0Z`,
	circle: (s) =>
		`M0 ${0.5 * s}` +
		`A${0.5 * s} ${0.5 * s} 0 0 0 ${s} ${0.5 * s}` +
		`A${0.5 * s} ${0.5 * s} 0 0 0 0 ${0.5 * s}`,
	diamond: (s) =>
		`M${0.5 * s} 0` +
		`A${0.6 * s} ${0.6 * s} 0 0 0 ${s} ${0.5 * s}` +
		`A${0.6 * s} ${0.6 * s} 0 0 0 ${0.5 * s} ${s}` +
		`A${0.6 * s} ${0.6 * s} 0 0 0 0 ${0.5 * s}` +
		`A${0.6 * s} ${0.6 * s} 0 0 0 ${0.5 * s} 0`,
	triangle: (s) =>
		`M${0.5 * s} ${0.0866 * s}L0 ${0.9234 * s}L${s} ${0.9234 * s}Z`,
	rhombus: (s) =>
		`M${0.5 * s} 0` +
		`L${s} ${0.5 * s}` +
		`L${0.5 * s} ${s}` +
		`L0 ${0.5 * s}Z`,
	star: (s) =>
		`M${0.5 * s} ${0.05 * s}` +
		`L${0.606 * s} ${0.36 * s}` +
		`L${s} ${0.36 * s}` +
		`L${0.685 * s} ${0.59 * s}` +
		`L${0.81 * s} ${0.95 * s}` +
		`L${0.5 * s} ${0.725 * s}` +
		`L${0.19 * s} ${0.95 * s}` +
		`L${0.315 * s} ${0.59 * s}` +
		`L0 ${0.36 * s}` +
		`L${0.394 * s} ${0.36 * s}Z`,
	heart: (s) =>
		`M${0.9 * s} ${0.5 * s}` +
		`A${0.08 * s} ${0.08 * s} 0 0 0 ${0.5 * s} ${0.2 * s}` +
		`A${0.08 * s} ${0.08 * s} 0 0 0 ${0.1 * s} ${0.5 * s}` +
		`L${0.5 * s} ${0.9 * s}` +
		`L${0.9 * s} ${0.5 * s}`,
	shurikan: (s) =>
		`M${0.3 * s} ${0.1 * s}L${0.5 * s} 0L${0.7 * s} ${0.1 * s}` +
		`A ${0.05 * s} ${0.05 * s} 0 0 0 ${0.9 * s} ${0.35 * s}` +
		`L${s} ${0.5 * s}L${0.9 * s} ${0.7 * s}` +
		`A ${0.05 * s} ${0.05 * s} 0 0 0 ${0.7 * s} ${0.9 * s}` +
		`L${0.5 * s} ${s}L${0.3 * s} ${0.9 * s}` +
		`A${0.05 * s} ${0.05 * s} 0 0 0 ${0.1 * s} ${0.7 * s}` +
		`L0 ${0.5 * s}L${0.1 * s} ${0.3 * s}` +
		`A${0.05 * s} ${0.05 * s} 0 0 0 ${0.3 * s} ${0.1 * s}` +
		`M${0.4 * s} ${0.5 * s}` +
		`A${0.1 * s} ${0.1 * s} 0 0 0 ${0.6 * s} ${0.5 * s}` +
		`A${0.1 * s} ${0.1 * s} 0 0 0 ${0.4 * s} ${0.5 * s}`,
	crosshair: (s) =>
		`M${0.2 * s} ${0.5 * s}` +
		`A${0.3 * s} ${0.3 * s} 0 0 0 ${0.8 * s} ${0.5 * s}` +
		`A${0.3 * s} ${0.3 * s} 0 0 0 ${0.2 * s} ${0.5 * s}` +
		`M0 ${0.5 * s}` +
		`L${s} ${0.5 * s}` +
		`M${0.5 * s} 0` +
		`L${0.5 * s} ${s}`,
	crossboats: (s) =>
		`M0 ${0.5 * s}` +
		`A${0.6 * s} ${0.4 * s} 0 0 0 ${s} ${0.5 * s}` +
		`A${0.6 * s} ${0.4 * s} 0 0 0 0 ${0.5 * s}` +
		`M${0.5 * s} 0` +
		`A${0.4 * s} ${0.6 * s} 0 0 0 ${0.5 * s} ${s}` +
		`A${0.4 * s} ${0.6 * s} 0 0 0 ${0.5 * s} 0`,
	curvedrhombus: (s) =>
		`M${0.1 * s} ${0.1 * s}` +
		`A${0.5 * s} ${0.5 * s} 0 0 0 ${0.9 * s} ${0.1 * s}` +
		`A${0.5 * s} ${0.5 * s} 0 0 0 ${0.9 * s} ${0.9 * s}` +
		`A${0.5 * s} ${0.5 * s} 0 0 0 ${0.1 * s} ${0.9 * s}` +
		`A${0.5 * s} ${0.5 * s} 0 0 0 ${0.1 * s} ${0.1 * s}`,
	fourflags: (s) =>
		`M${0.5 * s} ${0.3 * s}` +
		`A${0.2 * s} ${0.1 * s} 0 0 0 ${0.5 * s} ${0.1 * s}` +
		`L${0.5 * s} ${0.9 * s}` +
		`M${0.5 * s} ${0.7 * s}` +
		`A${0.2 * s} ${0.1 * s} 0 0 0 ${0.5 * s} ${0.9 * s}` +
		`M${0.3 * s} ${0.5 * s}` +
		`A${0.1 * s} ${0.2 * s} 0 0 0 ${0.1 * s} ${0.5 * s}` +
		`L${0.9 * s} ${0.5 * s}` +
		`M${0.7 * s} ${0.5 * s}` +
		`A${0.1 * s} ${0.2 * s} 0 0 0 ${0.9 * s} ${0.5 * s}`
}
