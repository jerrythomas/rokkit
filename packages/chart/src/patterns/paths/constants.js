import { scaledPathCollection } from '../../old_lib/utils'
import paths from './patterns.json'

export const patterns = scaledPathCollection(paths)
