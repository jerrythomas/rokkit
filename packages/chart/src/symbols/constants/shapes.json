{
  "circle": [
    ["M", 4.5, 0],
    ["A", 4.5, 4.5, 0, 1, 1, -4.5, 0],
    ["A", 4.5, 4.5, 0, 1, 1, 4.5, 0]
  ],
  "square": [
    ["M", 3.534, 3.534],
    ["L", 3.534, -3.534],
    ["L", -3.534, -3.534],
    ["L", -3.534, 3.534],
    ["Z"]
  ],
  "triangle": [["M", 0, -5.443], ["L", 4.714, 2.721], ["L", -4.714, 2.721], ["Z"]],
  "diamond": [["M", 0, -4.995], ["L", 4.995, 0], ["L", 0, 4.995], ["L", -4.995, 0], ["Z"]],
  "plus": [
    ["M", -6.873, 0],
    ["L", 6.873, 0],
    ["M", 0, 6.873],
    ["L", 0, -6.873]
  ],
  "asterisk": [
    ["M", 0, 4.769],
    ["L", 0, -4.769],
    ["M", -4.13, -2.384],
    ["L", 4.13, 2.384],
    ["M", -4.13, 2.384],
    ["L", 4.13, -2.384]
  ],
  "cross": [
    ["M", -4.87, -4.87],
    ["L", 4.87, 4.87],
    ["M", -4.87, 4.87],
    ["L", 4.87, -4.87]
  ],
  "heart": [
    ["M", 4, 0],
    ["A", 0.8, 0.8, 0, 0, 0, 0, -3],
    ["A", 0.8, 0.8, 0, 0, 0, -4, 0],
    ["L", 0, 4],
    ["L", 4, 0]
  ],
  "star": [
    ["M", -0.19999, -5],
    ["L", 0.92258, -1.54509],
    ["H", 4.55529],
    ["L", 1.61637, 0.59017],
    ["L", 2.73894, 4.04509],
    ["L", -0.19999, 1.90983],
    ["L", -3.13892, 4.04509],
    ["L", -2.01635, 0.59017],
    ["L", -4.95527, -1.54509],
    ["H", -1.32256],
    ["L", -0.19999, -5],
    ["Z"]
  ]
}
