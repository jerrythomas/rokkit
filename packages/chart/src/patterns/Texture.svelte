<script>
	import { library } from './constants'
	import templates from './templates'
	import { pick } from 'ramda'

	export let size = 10
	export let fill = 'none'
	export let stroke = 'none'
	export let name = 'Waves'

	$: pattern = library[name]
	$: props = { fill, stroke }
</script>

<svelte:component
	this={templates[pattern.component]}
	{size}
	{...pick(pattern.allowed, props)}
	data={pattern.data}
/>
