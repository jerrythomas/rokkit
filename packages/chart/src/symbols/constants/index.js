import { scaledPathCollection } from '../../old_lib/utils'
import paths from './shapes.json' with { type: 'json' }

export const namedShapes = scaledPathCollection(paths)
