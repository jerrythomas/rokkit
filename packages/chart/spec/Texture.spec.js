import { describe, it, expect } from 'vitest'
// skipcq: JS-C1003 - Importing all components for verification
import Texture from '../src/template/Texture.svelte'
import { library } from '../src/template/constants'
import { render } from '@testing-library/svelte'
import * as templates from '../src/template/shapes'

describe('templates', () => {
	it('should contain all exported components', () => {
		expect(Object.keys(templates)).toEqual(['Circles', 'Lines', 'Path', 'Polygons'])
	})
})

describe('Texture.svelte', () => {
	it.each(Object.keys(library))('should render pattern "%s" using Texture', (key) => {
		const { container } = render(Texture, {
			props: { name: key }
		})
		expect(container).toMatchSnapshot()
	})
})
