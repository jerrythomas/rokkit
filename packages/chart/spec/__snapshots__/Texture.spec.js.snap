// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Texture.svelte > should render pattern "Brick" using Texture 1`] = `
<div>
  <!---->
  <!---->
  <line
    stroke="none"
    stroke-linecap="round"
    stroke-width="0.5"
    x1="0"
    x2="5"
    y1="2.5"
    y2="2.5"
  />
  <line
    stroke="none"
    stroke-linecap="round"
    stroke-width="0.5"
    x1="5"
    x2="10"
    y1="7.5"
    y2="7.5"
  />
  <line
    stroke="none"
    stroke-linecap="round"
    stroke-width="0.5"
    x1="0"
    x2="0"
    y1="0"
    y2="10"
  />
  <line
    stroke="none"
    stroke-linecap="round"
    stroke-width="0.5"
    x1="10"
    x2="10"
    y1="0"
    y2="10"
  />
  <line
    stroke="none"
    stroke-linecap="round"
    stroke-width="0.5"
    x1="5"
    x2="5"
    y1="0"
    y2="10"
  />
  
  
  
</div>
`;

exports[`Texture.svelte > should render pattern "Circles" using Texture 1`] = `
<div>
  <!---->
  <!---->
  <circle
    cx="0"
    cy="0"
    fill="none"
    r="5"
    stroke="none"
  />
  <circle
    cx="10"
    cy="10"
    fill="none"
    r="5"
    stroke="none"
  />
  
  
  
</div>
`;

exports[`Texture.svelte > should render pattern "Dominoes" using Texture 1`] = `
<div>
  <!---->
  <!---->
  <circle
    cx="2"
    cy="2"
    fill="none"
    r="8"
    stroke="currentColor"
  />
  <circle
    cx="4"
    cy="4"
    fill="none"
    r="8"
    stroke="currentColor"
  />
  <circle
    cx="6"
    cy="6"
    fill="none"
    r="8"
    stroke="currentColor"
  />
  <circle
    cx="8"
    cy="8"
    fill="none"
    r="8"
    stroke="currentColor"
  />
  <circle
    cx="8"
    cy="2"
    fill="none"
    r="8"
    stroke="currentColor"
  />
  <circle
    cx="6"
    cy="4"
    fill="none"
    r="8"
    stroke="currentColor"
  />
  <circle
    cx="4"
    cy="6"
    fill="none"
    r="8"
    stroke="currentColor"
  />
  <circle
    cx="2"
    cy="8"
    fill="none"
    r="8"
    stroke="currentColor"
  />
  
  
  
</div>
`;

exports[`Texture.svelte > should render pattern "Waves" using Texture 1`] = `
<div>
  <!---->
  <path
    d="M 0 2.5 A 6 5 0 0 0 10 2.5"
    fill="none"
    stroke="currentColor"
    stroke-width="0.5"
  />
  
  
</div>
`;
